rllib/env/remote_base_env.py

```diff
@@ -0,0 +1,225 @@
+import logging
+from typing import Tuple, Callable, List, Optional
+
+import ray
+from ray.rllib.env.base_env import BaseEnv, _DUMMY_AGENT_ID, ASYNC_RESET_RETURN
+from ray.rllib.utils.annotations import override, PublicAPI
+from ray.rllib.utils.typing import MultiEnvDict, EnvType, EnvID, MultiAgentDict
+
+logger = logging.getLogger(__name__)
+
+
+@PublicAPI
+class RemoteBaseEnv(BaseEnv):
+    """BaseEnv that executes its sub environments as @ray.remote actors.
+
+    This provides dynamic batching of inference as observations are returned
+    from the remote simulator actors. Both single and multi-agent child envs
+    are supported, and envs can be stepped synchronously or asynchronously.
+
+    You shouldn't need to instantiate this class directly. It's automatically
+    inserted when you use the `remote_worker_envs=True` option in your
+    Trainer's config.
+    """
+
+    def __init__(self,
+                 make_env: Callable[[int], EnvType],
+                 num_envs: int,
+                 multiagent: bool,
+                 remote_env_batch_wait_ms: int,
+                 existing_envs: Optional[List[ray.actor.ActorHandle]] = None):
+        """Initializes a RemoteVectorEnv instance.
+
+        Args:
+            make_env: Callable that produces a single (non-vectorized) env,
+                given the vector env index as only arg.
+            num_envs: The number of sub-environments to create for the
+                vectorization.
+            multiagent: Whether this is a multiagent env or not.
+            remote_env_batch_wait_ms: Time to wait for (ray.remote)
+                sub-environments to have new observations available when
+                polled. When none of the sub-environments is ready, simply
+                repeat the ray.wait call until at least one sub-env is
+                ready.
+            existing_envs: Optional list of already created sub-environments.
+                These will be used as-is and only as many new sub-envs as
+                necessary (`num_envs - len(existing_envs)`) will be created.
+        """
+
+        # Could be creating local or remote envs.
+        self.make_env = make_env
+        # Whether the given `make_env` callable already returns ray.remote
+        # objects or not.
+        self.make_env_creates_actors = False
+        # Already existing env objects (generated by the RolloutWorker).
+        self.existing_envs = existing_envs or []
+        self.num_envs = num_envs
+        self.multiagent = multiagent
+        self.poll_timeout = remote_env_batch_wait_ms / 1000
+
+        self.actors = None  # lazy init
+        self.pending = None  # lazy init
```

Comment:
any chance you can explain the role of self.pending a bit here?
I found it to be quite important, carrying logic between member functions actually.

Reply:
Added a comment.